### DISCUSSION TITLE
"Wrong boot param version 8, need 9"

### DISCUSSION THREAD

**Discussion by EricOulashin** [dosemu2/dosemu2#1931] [Q&A]
Today I updated dosemu2 to 2.0~pre9-7947, and now when I run something, dosemu2 outputs "Wrong boot param version 8, need 9". It still continues on and runs the program, so this isn't a big issue, but I'm not sure what that message means.  Is there a different command-line parameter I need to use now?

I'm using dosemu2 primarily to run DOS-based BBS door games & things on my BBS.  I see that message when I run one, just before it continues on to run the game.

**Comment by stsp** [comment]
This is likely a temporary problem,
some components updated before
others. Try updating again tomorrow
and that should go away.

**Comment by stsp** [comment]
I triggered a rebuild on PPA to make
sure the problem doesn't stick.
So try updating eg half an hour later
or so.

**Reply by EricOulashin** [reply]
I've installed the latest update, and the error is no longer showing. Thanks!

**Comment by emendelson** [comment]
Running on fully updated Ubuntu Mate, with updated dosemu and fdpp, I'm getting a similar error message, but with `Wrong boot param 7, need 9`.

`ERROR: fdpp: PANIC: Wrong boot param version 7, need 9`

dosemu2 version number is `2.0~pre9-7948-a40569614+202303022044~ubuntu22.04.1`

fdpp version number is `1.6-1334-91a6ee1+202303031246~ubuntu22.04.1`

I tried removing and reinstalling dosemu2, fdpp, and comcom32, but it didn't help. I'll try again tomorrow, but I thought it might be worth mentioning.

What is very strange is that the same versions do NOT produce that error message on a non-Mate Ubuntu setup.

**Reply by stsp** [reply]
This is very wrong.
I think you have some local dosemu
build somewhere that interferes.
Remember you tested the patch?
You probably did `sudo make install`.
Try `sudo make uninstall`.

**Reply by emendelson** [reply]
Yes, that fixed it. Thank you for reminding me.

**Reply by stsp** [reply]
That hints about PATH problems on your
side. You probably had dosemu from
/usr/local executing in preference to /usr.

**Reply by emendelson** [reply]
My PATH (unchanged by me; it's what got installed with Ubuntu Mate) is this:

`/home/edward:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin`

Your dosemu is in `/usr/bin`; mine seems to have been in `/usr/local/bin`, exactly as you said.

**Reply by stsp** [reply]
Right, seems all distros have moved to
putting /usr/local/bin before /usr/bin, which
is IIRC wasn't the case in the past.
I find such a move strange, as FHS states
that /usr/local is for
`programs and data that are shareable amongst a group of hosts, but not found in /usr.`
So it almost explicitly says that /usr is
searched before...
Confused.
Either I am reading FHS wrongly, or distros
do.

**Reply by emendelson** [reply]
Possibly the idea is that the user has built and installed a program (as I did), and therefore prefers to use the user's own build, instead of opening the standard one? That makes at least some kind of sense.

**Reply by stsp** [reply]
That seems to contradict what
FHS says, and overall FHS makes
sense to me, as it is absolutely not
good to have different versions of
the same program installed in
multiple global places. That will
lead to problems, such as what
we've seen here. So FHS clearly
implies you just need to uninstall
the software from /usr first, and
then it will work in /usr/local.

If you want to override the globally
installed software, you need to use
the local store, like `~/.local`. But
that seems to be not standardized
by FHS (at least I can't find it there).
So to me it seems, its a mess where
distros decided to ignore FHS and
do what they want.

I also violated FHS (didn't know back
then) when moved .dosemurc to
`~/.dosemu`. FHS clearly says the
leading dot had to be removed when
doing such a move...

**Reply by stsp** [reply]
Additional complication here is that
`PKG_CHECK_MODULES()` autoconf
macro doesn't take prefix into an account.
So it would find fdpp installed in /usr
even if you build dosemu2 for working
in /usr/local, and there seems to be no
way of changing that behavior. Otherwise
we could force the user to install fdpp
into /usr/local manually, and then at
least things would stay in sync. But if
the prefix is not taken into account, then
the only way to get the consistent system,
is to remove the software from /usr before
adding it to /usr/local. This is what FHS
seems to imply (in my reading of it), and
therefore I am sure the PATH should have
/usr/loca/bin at the end, not at the beginning.

**Reply by skitt** [reply]
I’ve seen `/usr/local/bin` ahead of `/bin` and `/usr/bin` on Unix-style systems since before Linux, this isn’t a recent development. I just checked, Slackware 3.0 sets up a default path of `/usr/local/bin:/bin:/usr/bin` for users (`/usr/local/sbin:/usr/sbin:/sbin:/usr/local/bin:/bin:/usr/bin` for root), and IIRC similar-vintage Debian and Red Hat distributions did the same. Of course it’s best to avoid confusion by not having same-name binaries in `/usr/bin` and `/usr/local/bin`, but that’s not always possible.

**Reply by stsp** [reply]
> I’ve seen /usr/local/bin ahead of /bin and /usr/bin on Unix-style systems since before Linux

I've booted a freebsd-13.1 VM to check that,
as that your assertion would mean my memory
is a bit corrupted.
```
# echo $PATH
/sbin:/bin:/usr/sbin:/usr/bin:/usr/local/sbin:/usr/local/bin:/root/bin
```
Only that layout makes sense to me and
seems to conform to FHS.
As for linux-based distros, they seem to
have been fiddling with PATH too:
https://fedoraproject.org/wiki/Changes/UserPathPrioritization
But that change seems to be not
directly related to /usr/local.

**Reply by skitt** [reply]
I’m not claiming it’s _universally_ ahead, just that I’ve seen it that way for a long time and it’s not a recent change in Linux distributions. See for example `pam_env.conf` (which is what Fedora and RHEL use to configure the default `PATH`): it had `/usr/local/bin` ahead of `/usr/bin` etc. [in 2000](https://github.com/linux-pam/linux-pam/blob/ea488580c42e8918445a945484de3c8a5addc761/modules/pam_env/pam_env.conf-example#L64) and perhaps even earlier.

**Reply by stsp** [reply]
> I’m not claiming it’s universally ahead

I didn't mean you claim that.
What you said, was
```
I’ve seen /usr/local/bin ahead of /bin and /usr/bin on Unix-style systems since before Linux
```
and so I checked freebsd.
Of course I can't make a too wide claims,
but I hope you haven't seen that on
"Unix-style systems since before Linux",
or if you did - they were not very popular. :)

**Reply by skitt** [reply]
SunOS ;-)

**Comment by stsp** [comment]
OK I added soname to fdpp.
That should disallow any possible
usage of the wrong libs.
But still its not a full solution, for
example its not a solution for bisecting.
Not sure how to create all non-latest
versions of fdpp to provide them in
parallel, allowing any dosemu version
to work...
